File: agents/issue-reconciler/api/fetchers/candidate-prs.ts
import type { GitHubClient } from '../client.ts';
import { FUZZY_CANDIDATE_LIMIT, FUZZY_CANDIDATE_LOOKBACK_DAYS, REPO_NAME, REPO_OWNER } from '../../config.ts';

export type CandidatePR = {
  number: number;
  title: string;
  headRefName: string;
  state: 'OPEN' | 'MERGED';
  merged: boolean;
  mergedAt: string | null;
  isDraft: boolean;
};

const PR_FIELDS = `number title headRefName state isDraft mergedAt`;

const QUERY = `
  query CandidatePRs($owner: String!, $name: String!, $limit: Int!) {
    repository(owner: $owner, name: $name) {
      open: pullRequests(states: [OPEN], first: $limit, orderBy: { field: UPDATED_AT, direction: DESC }) {
        nodes { ${PR_FIELDS} }
      }
      merged: pullRequests(states: [MERGED], first: $limit, orderBy: { field: UPDATED_AT, direction: DESC }) {
        nodes { ${PR_FIELDS} }
      }
    }
  }
`;

type RawPR = { number: number; title: string; headRefName: string; state: 'OPEN' | 'MERGED'; isDraft: boolean; mergedAt: string | null };

type RawResponse = {
  repository: {
    open: { nodes: RawPR[] };
    merged: { nodes: RawPR[] };
  } | null;
};

function toCandidatePR(pr: RawPR): CandidatePR {
  return {
    number: pr.number,
    title: pr.title,
    headRefName: pr.headRefName,
    state: pr.state,
    merged: pr.state === 'MERGED',
    mergedAt: pr.mergedAt,
    isDraft: pr.isDraft,
  };
}

/**
 * A bounded, repo-wide pool of open PRs plus PRs merged in the last
 * FUZZY_CANDIDATE_LOOKBACK_DAYS days - one query per run, not per issue. This
 * is the search space the fuzzy matcher (matcher/fuzzy.ts) compares each
 * unmatched issue against; it is not itself evidence for any one issue.
 * Carries full PR state (not just title/branch) so a fuzzy match to a merged
 * PR is distinguishable from a match to one still open.
 */
export async function fetchCandidatePRs(client: GitHubClient, now: Date = new Date()): Promise<CandidatePR[]> {
  const data = await client.query<RawResponse>(QUERY, {
    owner: REPO_OWNER,
    name: REPO_NAME,
    limit: FUZZY_CANDIDATE_LIMIT,
  });

  const repository = data.repository;
  if (!repository) return [];

  const cutoff = now.getTime() - FUZZY_CANDIDATE_LOOKBACK_DAYS * 24 * 60 * 60 * 1000;
  const recentlyMerged = repository.merged.nodes.filter(
    (pr) => pr.mergedAt !== null && new Date(pr.mergedAt).getTime() >= cutoff,
  );

  const candidates = [...repository.open.nodes, ...recentlyMerged].map(toCandidatePR);
  const seen = new Set<number>();
  return candidates.filter((pr) => {
    if (seen.has(pr.number)) return false;
    seen.add(pr.number);
    return true;
  });
}

File: agents/issue-reconciler/issue_reconciler/api/fetchers/candidate_prs.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import TypedDict

from issue_reconciler.api.client import GitHubClient
from issue_reconciler.config import (
    FUZZY_CANDIDATE_LIMIT,
    FUZZY_CANDIDATE_LOOKBACK_DAYS,
    REPO_NAME,
    REPO_OWNER,
)

_PR_FIELDS = "number title headRefName state isDraft mergedAt"

_QUERY = f"""
  query CandidatePRs($owner: String!, $name: String!, $limit: Int!) {{
    repository(owner: $owner, name: $name) {{
      open: pullRequests(states: [OPEN], first: $limit, orderBy: {{ field: UPDATED_AT, direction: DESC }}) {{
        nodes {{ {_PR_FIELDS} }}
      }}
      merged: pullRequests(states: [MERGED], first: $limit, orderBy: {{ field: UPDATED_AT, direction: DESC }}) {{
        nodes {{ {_PR_FIELDS} }}
      }}
    }}
  }}
"""


class CandidatePR(TypedDict):
    number: int
    title: str
    head_ref_name: str
    state: str  # "OPEN" | "MERGED"
    merged: bool
    merged_at: str | None
    is_draft: bool


def _to_candidate(pr: dict) -> CandidatePR:
    return {
        "number": pr["number"],
        "title": pr["title"],
        "head_ref_name": pr["headRefName"],
        "state": pr["state"],
        "merged": pr["state"] == "MERGED",
        "merged_at": pr.get("mergedAt"),
        "is_draft": pr["isDraft"],
    }


def fetch_candidate_prs(client: GitHubClient, now: datetime | None = None) -> list[CandidatePR]:
    """A bounded, repo-wide pool of open PRs plus PRs merged in the last
    FUZZY_CANDIDATE_LOOKBACK_DAYS days - one query per run, not per issue.
    This is the search space the fuzzy matcher compares each unmatched issue
    against; it is not itself evidence for any one issue. Carries full PR
    state (not just title/branch) so a fuzzy match to a merged PR is
    distinguishable from a match to one still open.
    """
    now = now or datetime.now(timezone.utc)
    data = client.query(_QUERY, {"owner": REPO_OWNER, "name": REPO_NAME, "limit": FUZZY_CANDIDATE_LIMIT})
    repository = data.get("repository")
    if not repository:
        return []

    cutoff = now - timedelta(days=FUZZY_CANDIDATE_LOOKBACK_DAYS)
    recently_merged = [
        pr for pr in repository["merged"]["nodes"] if pr.get("mergedAt") and datetime.fromisoformat(pr["mergedAt"]) >= cutoff
    ]

    seen: set[int] = set()
    candidates: list[CandidatePR] = []
    for pr in [*repository["open"]["nodes"], *recently_merged]:
        candidate = _to_candidate(pr)
        if candidate["number"] in seen:
            continue
        seen.add(candidate["number"])
        candidates.append(candidate)

    return candidates
